src/linear_algebra/free_module.lean

```diff
@@ -131,17 +135,89 @@ instance pi {ι : Type*} [fintype ι] {M : ι → Type*} [Π (i : ι), add_comm_
 [Π (i : ι), module R (M i)] [Π (i : ι), module.free R (M i)] : module.free R (Π i, M i) :=
 of_basis $ pi.basis $ λ i, choose_basis R (M i)
 
+instance matrix {ι₁ : Type*} [fintype ι₁] {ι₂ : Type*} [fintype ι₂] :
+  module.free R (matrix ι₁ ι₂ R) :=
+of_basis $ pi.basis $ λ i, pi.basis_fun _ _
+
 end semiring
 
+section ring
+
+variables [ring R] [add_comm_group M] [module R M] [module.free R M]
+
+/-- If a free module is finite, then any basis is finite. -/
+noncomputable
+instance [nontrivial R] [module.finite R M] :
+  fintype (module.free.choose_basis_index R M) :=
```

Comment:
@semorrison Did you recent work on linear algebra give us tools to golf this proof?

Reply:
Yes, I think this was done as
```
def basis_fintype_of_finite_spans (w : set M) [fintype w] (s : span R w = ⊤)
  {ι : Type w} (b : basis ι R M) : fintype ι :=
```
which works for any nontrivial ring.

Reply:
Thank you! I wonder if this is true over a semiring...